src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -1060,12 +1203,61 @@ struct BlockZipperJoinImpl {
   }
 
   // The actual join routine that combines all the previous functions.
+  // not been consumed yet and match them against undef blocks.
+  template <bool reversed>
+  bool consumeRemainingBlocks(auto& side, const auto& undefBlocks) {
+    bool hasUndef = false;
+    while (side.it_ != side.end_) {
+      const auto& lBlock = *side.it_;
+      for (const auto& rBlock : undefBlocks) {
+        if constexpr (reversed) {
+          compatibleRowAction_.setInput(rBlock.fullBlock(), lBlock);
+        } else {
+          compatibleRowAction_.setInput(lBlock, rBlock.fullBlock());
+        }
+        for (size_t i : ad_utility::integerRange(lBlock.size())) {
+          for (size_t j : rBlock.getIndexRange()) {
+            hasUndef = true;
+            if constexpr (reversed) {
+              compatibleRowAction_.addRow(j, i);
+            } else {
+              compatibleRowAction_.addRow(i, j);
+            }
+          }
+        }
+      }
+      ++side.it_;
+    }
```

Comment:
This seems very related to the various `cartesian products` that we already have above, the the sides seem to have different values. 
Can you make the code more symmetric by packing the "naked" side into the `BlockAndSubrangeOrWhatever` wrapper, s.t. you can reuse more code here and can make the `reverse` case easier?

Reply:
I can't quite follow you there. You'll have to explain your proposal a little bit clearer.